query_execution/QueryExecutionState.hpp

```diff
@@ -0,0 +1,294 @@
+/**
+ *   Copyright 2016, Quickstep Research Group, Computer Sciences Department,
+ *     University of Wisconsin—Madison.
+ *
+ *   Licensed under the Apache License, Version 2.0 (the "License");
+ *   you may not use this file except in compliance with the License.
+ *   You may obtain a copy of the License at
+ *
+ *       http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *   Unless required by applicable law or agreed to in writing, software
+ *   distributed under the License is distributed on an "AS IS" BASIS,
+ *   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *   See the License for the specific language governing permissions and
+ *   limitations under the License.
+ **/
+
+#ifndef QUICKSTEP_QUERY_EXECUTION_QUERY_EXECUTION_STATE_HPP_
+#define QUICKSTEP_QUERY_EXECUTION_QUERY_EXECUTION_STATE_HPP_
+
+#include <cstddef>
+#include <unordered_map>
+#include <utility>
+#include <vector>
+
+#include "utility/Macros.hpp"
+
+#include "glog/logging.h"
+
+namespace quickstep {
+
+/** \addtogroup QueryExecution
+ *  @{
+ */
+
+/**
+ * @brief A class that tracks the state of the execution of a query which
+ *        includes status of operators, number of dispatched work orders etc.
+ **/
+class QueryExecutionState {
+ public:
+  /**
+   * @brief Constructor.
+   *
+   * @param num_operators Number of relational operators in the query.
+   **/
+  explicit QueryExecutionState(const std::size_t num_operators)
+      : num_operators_(num_operators), num_operators_finished_(0) {
+    done_gen_.assign(num_operators, false);
+    execution_finished_.assign(num_operators, false);
+    rebuild_required_.assign(num_operators, false);
+
+    queued_workorders_per_op_.assign(num_operators, 0);
```

Comment:
Refactor using the init list:

```
  ...
  num_operators_finished_(0),
  queued_workorders_per_op_(num_operators, 0),
  rebuild_required_(num_operators, false),
  execution_finished_(num_operators, false) {}  
```

Note that you have to make sure the init order be the same as the declaration in the private section..